lib/nostrum/shard/session.ex

```diff
@@ -32,45 +37,65 @@ defmodule Nostrum.Shard.Session do
 
   def request_guild_members(pid, guild_id, limit \\ 0) do
     payload = Payload.request_members_payload(guild_id, limit)
-    WebSockex.cast(pid, {:request_guild_members, payload})
+    GenServer.cast(pid, {:request_guild_members, payload})
   end
 
   def start_link([gateway, shard_num]) do
+    GenServer.start_link(__MODULE__, [gateway, shard_num], spawn_opt: [Util.fullsweep_after()])
+  end
+
+  def init([_gateway, _shard_num] = args) do
+    {:ok, nil, {:continue, args}}
+  end
+
+  def handle_continue([gateway, shard_num], nil) do
+    Connector.block_until_connect()
+    Logger.metadata(shard: shard_num)
+
+    {:ok, worker} = :gun.open(:binary.bin_to_list(gateway), 443, %{protocols: [:http]})
+    {:ok, :http} = :gun.await_up(worker, @timeout_connect)
+    stream = :gun.ws_upgrade(worker, @gateway_qs)
+    await_ws_upgraded(worker, stream)
+
+    zlib_context = :zlib.open()
+    :zlib.inflateInit(zlib_context)
+
     state = %WSState{
+      conn_pid: self(),
+      conn: worker,
       shard_num: shard_num,
       gateway: gateway <> @gateway_qs,
       last_heartbeat_ack: DateTime.utc_now(),
-      heartbeat_ack: true
+      heartbeat_ack: true,
+      zlib_ctx: zlib_context
     }
 
-    Connector.block_until_connect()
+    Logger.debug(fn -> "Websocket connection up on worker #{inspect(worker)}." end)
 
-    # TODO: Add support for `spawn_opt` start arguments to WebSockex, this does nothing until then.
-    WebSockex.start_link(
-      gateway <> @gateway_qs,
-      __MODULE__,
-      state,
-      spawn_opt: [Util.fullsweep_after()]
-    )
+    {:noreply, state}
   end
 
-  def handle_connect(conn, state) do
-    Logger.metadata(shard: state.shard_num)
+  defp await_ws_upgraded(worker, stream) do
```

Comment:
Can we change this to be more present tense and less past tense?

Reply:
Yes